bootstrap/src/main/java/io/airlift/bootstrap/LifeCycleManager.java

```diff
@@ -221,14 +238,49 @@ public void addInstance(Object instance)
         }
     }
 
+    private void stopInstance(Object obj, LifeCycleStopFailureHandler handler)
+    {
+        log.debug("Stopping %s", obj.getClass().getName());
+        LifeCycleMethods methods = methodsMap.get(obj.getClass());
+        for (Method preDestroy : methods.methodsFor(PreDestroy.class)) {
+            log.debug("\t%s()", preDestroy.getName());
+            try {
+                preDestroy.invoke(obj);
+            }
+            catch (Exception e) {
+                handler.handlePreDestroyException(obj.getClass(), preDestroy, unwrapInvocationTargetException(e));
+            }
+        }
+    }
+
     private void startInstance(Object obj)
-            throws IllegalAccessException, InvocationTargetException
+            throws Exception
     {
         log.debug("Starting %s", obj.getClass().getName());
         LifeCycleMethods methods = methodsMap.get(obj.getClass());
         for (Method postConstruct : methods.methodsFor(PostConstruct.class)) {
             log.debug("\t%s()", postConstruct.getName());
-            postConstruct.invoke(obj);
+            try {
+                postConstruct.invoke(obj);
+            }
+            catch (Exception e) {
+                Exception cause = unwrapInvocationTargetException(e);
+                log.debug(cause, "Exception in PostConstruct method %s::%s()", obj.getClass().getName(), postConstruct.getName());
+                stopInstance(obj, (Class<?> klass, Method method, Exception exception) -> {
+                    log.debug(exception, "Exception in PreDestroy method %1$s::%2$s() after PostConstruct failure in %1$s::%3$s()", klass.getName(), method.getName(), postConstruct.getName());
+                    cause.addSuppressed(exception);
+                });
+                throw cause;
+            }
         }
     }
+
+    /**
+     * Unwraps {@link InvocationTargetException} instances to their underlying cause, otherwise returns the provided
+     * Exception
```

Comment:
Nit: we don't have a hard line length limit, so either put this all on one line, or balance the text onto two lines (having a single word on a line looks strange)